src/AaveV3StrategicAssets_20230606/VeTokenManager.sol

```diff
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.19;
+
+import {IERC20} from 'solidity-utils/contracts/oz-common/interfaces/IERC20.sol';
+import {SafeERC20} from 'solidity-utils/contracts/oz-common/SafeERC20.sol';
+
+import {DelegateRegistry} from './interfaces/IDelegateRegistry.sol';
+import {IVeToken} from './interfaces/IVeToken.sol';
+import {IWardenBoost} from './interfaces/IWardenBoost.sol';
+import {IAragonVoting} from './interfaces/IAragonVoting.sol';
+import {Core} from './Core.sol';
+
+abstract contract VeTokenManager is Core {
+  using SafeERC20 for IERC20;
+
+  event BuyBoost(address delegator, address receiver, uint256 amount, uint256 duration);
+  event DelegateUpdate(address indexed oldDelegate, address indexed newDelegate);
+  event Lock(uint256 cummulativeTokensLocked, uint256 lockHorizon);
+  event Unlock(uint256 tokensUnlocked);
+  event VoteCast(uint256 voteData, bool support);
+  event VotingContractUpdate(address indexed token, address voting);
+
+  /// @notice The new lock duration must be greater than the current one
+  error InvalidNewDuration();
+  /// @notice The on-chain voting contract has not been set for this token
+  error VotingContractNotRegistered();
+
+  struct VeToken {
+    /// @notice the keccak encoded spaceId of the snapshot space
+    bytes32 spaceId;
+    /// @notice The lock duration of veToken
+    uint256 lockDuration;
+    /// @notice Address of veToken
+    address veToken;
+    /// @notice the snapshot delegate for the deposit
+    address delegate;
+    /// @notice Warden contract address for veBoost
+    address warden;
+  }
+
+  DelegateRegistry public constant DELEGATE_REGISTRY =
+    DelegateRegistry(0x469788fE6E9E9681C6ebF3bF78e7Fd26Fc015446);
+
+  mapping(address => VeToken) public veTokens;
+  mapping(address => address) public veTokenToVotingContract;
+
+  /// @notice The Buyer will then need to call the user_checkpoint() method on all Curve V4 & Factory Gauges
+  /// currently farming, to apply the newly received Boost.
+  /// @notice duration in weeks (ie: 1 for 1 Week)
+  function buyBoost(
+    address underlying,
+    address delegator,
+    address receiver,
+    uint256 amount,
+    uint256 duration
+  ) external onlyOwnerOrManager {
+    VeToken storage token = veTokens[underlying];
+    if (token.warden == address(0)) revert TokenNotRegistered();
+
+    IERC20 feeToken = IWardenBoost(token.warden).feeToken();
+    uint256 maxFee = IWardenBoost(token.warden).estimateFees(delegator, amount, duration);
+    feeToken.safeApprove(token.warden, maxFee);
```

Comment:
why safeApprove?
If we buy boost twice for the same warden, and during the first call if  `buyDelegationBoost()` consumes gas lesser than the maxGas the allowance of this contract to the warden will be non-zero will revert right?